dropwizard-auth/src/main/java/io/dropwizard/auth/CachingAuthenticator.java

```diff
@@ -51,25 +54,37 @@ public CachingAuthenticator(final MetricRegistry metricRegistry,
     public CachingAuthenticator(final MetricRegistry metricRegistry,
                                 final Authenticator<C, P> authenticator,
                                 final CacheBuilder<Object, Object> builder) {
-        this.underlying = authenticator;
         this.cacheMisses = metricRegistry.meter(name(authenticator.getClass(), "cache-misses"));
         this.gets = metricRegistry.timer(name(authenticator.getClass(), "gets"));
-        this.cache = builder.recordStats().build();
+        this.cache = builder.recordStats().build(new CacheLoader<C, Optional<P>>() {
+            @Override
+            public Optional<P> load(C key) throws Exception {
+                cacheMisses.mark();
+                final Optional<P> optPrincipal = authenticator.authenticate(key);
+                if (!optPrincipal.isPresent()) {
+                    // Prevent caching of unknown credentials
+                    throw new InvalidCredentialsException();
+                }
+                return optPrincipal;
+            }
+        });
     }
 
     @Override
     public Optional<P> authenticate(C credentials) throws AuthenticationException {
         final Timer.Context context = gets.time();
         try {
-            Optional<P> optionalPrincipal = cache.getIfPresent(credentials);
-            if (optionalPrincipal == null) {
-                cacheMisses.mark();
-                optionalPrincipal = underlying.authenticate(credentials);
-                if (optionalPrincipal.isPresent()) {
-                    cache.put(credentials, optionalPrincipal);
-                }
+            return cache.get(credentials);
+        } catch (UncheckedExecutionException|ExecutionException e) {
+            final Throwable cause = e.getCause();
+            if (cause instanceof InvalidCredentialsException) {
+                return Optional.empty();
             }
-            return optionalPrincipal;
+            // Attempt to re-throw as-is
+            Throwables.propagateIfPossible(cause, AuthenticationException.class);
+
+            // Rethrow other checked exceptions wrapped in an AuthenticationException (this shouldn't happen)
+            throw new AuthenticationException(Optional.ofNullable(cause).orElse(e));
```

Comment:
Despite it's possible to have a null value for a cause for an exception, I think we can't get an `ExecutionException` without a cause from a `LoadingCache`, so it's safe to avoid this check.